File: firebase/functions/lib/admin/payments/get_stripe_billing_portal_link.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import 'stripe_util.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';

class GetStripeBillingPortalLink
    extends OnCallMethod<GetStripeBillingPortalLinkRequest> {
  GetStripeBillingPortalLink()
      : super(
          'getStripeBillingPortalLink',
          (json) => GetStripeBillingPortalLinkRequest.fromJson(json),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetStripeBillingPortalLinkRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final stripeCustomerId =
        await stripeUtil.getOrCreateCustomerStripeId(uid: context.authUid!);

    final domain = functions.config.get('app.domain') as String;

    final Map<String, String> params = {
      'customer': stripeCustomerId,
      'return_url': 'https://$domain/${request.responsePath}',
    };

    final jsonResponse =
        await stripeUtil.post(path: '/billing_portal/sessions', params: params);
    return GetStripeBillingPortalLinkResponse(url: jsonResponse['url'])
        .toJson();
  }

  // Sets minimum instances to 0
  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 0,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/admin/payments/get_stripe_billing_portal_link.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { stripeUtil } from './stripe_util';
import { orElseUnauthorized } from '../../utils/utils';

interface GetStripeBillingPortalLinkRequest {
  responsePath: string;
}

export class GetStripeBillingPortalLink extends OnCallMethod<GetStripeBillingPortalLinkRequest> {
  constructor() {
    super(
      'getStripeBillingPortalLink',
      (json) => json as GetStripeBillingPortalLinkRequest,
      { runWithOptions: { timeoutSeconds: 60, memory: '1GB', minInstances: 0 } }
    );
  }

  async action(request: GetStripeBillingPortalLinkRequest, context: functions.https.CallableContext): Promise<Record<string, unknown>> {
    orElseUnauthorized(context.auth?.uid != null);

    const stripeCustomerId = await stripeUtil.getOrCreateCustomerStripeId({ uid: context.auth!.uid });
    const domain = functions.config().app?.domain as string;

    const params: Record<string, string> = {
      'customer': stripeCustomerId,
      'return_url': `https://${domain}/${request.responsePath}`,
    };

    const jsonResponse = await stripeUtil.post({ path: '/billing_portal/sessions', params });
    return { url: jsonResponse['url'] };
  }
}
